DXMainClient/DXGUI/Multiplayer/CnCNet/GlobalContextMenu.cs

```diff
@@ -153,24 +142,77 @@ private void UpdatePlayerBasedButtons()
+                return;
+            }
 
-                ProcessLauncher.StartShellProcess(link);
-            };
+            ClientRectangle = LNK_SIZE;
+
+            foreach (string link in links)
+            {
+                string linkToDisplay = LINK_LENGTH > link.Length ? link.Substring(0, link.Length) : link.Substring(0, LINK_LENGTH) + "...";
+
+                if (Items.Where(item => item.Text.Contains(linkToDisplay)).ToList().Count > 0)
+                    continue;
+
+                var copyLinkItem = new XNAContextMenuItem()
+                {
+                    Text = string.Format("{0} {1}", COPY_LINK, linkToDisplay),
+                    SelectAction = () => CopyLink(link)
+                };
+
+                var openLinkItem = new XNAContextMenuItem()
+                {
+                    Text = string.Format("{0} {1}", OPEN_LINK, linkToDisplay),
+                    SelectAction = () =>
+                    {
+                        bool isTrusted = false;
+                        try
+                        {
+                            string domain = new Uri(link).Host;
+                            var trustedDomains = ClientConfiguration.Instance.TrustedDomains.Concat(ClientConfiguration.Instance.AlwaysTrustedDomains);
+                            isTrusted = trustedDomains.Contains(domain, StringComparer.InvariantCultureIgnoreCase)
+                                || trustedDomains.Any(trustedDomain => domain.EndsWith("." + trustedDomain, StringComparison.InvariantCultureIgnoreCase));
+                        }
+                        catch (Exception ex)
+                        {
+                            isTrusted = false;
+                            Logger.Log($"Error in parsing the URL \"{link}\": {ex.ToString()}");
+                        }
+
+                        if (isTrusted)
+                        {
+                            ProcessLauncher.StartShellProcess(link);
+                            return;
+                        }
+
+                        // Show the warning if the links is not trusted
+                        var msgBox = new XNAMessageBox(WindowManager,
+                            "Open Link Confirmation".L10N("Client:Main:OpenLinkConfirmationTitle"),
+                            """
+                            You're about to open a link shared in chat.
+
+                            Please note that this link hasn't been verified,
+                            and CnCNet is not responsible for its content.
+
+                            Would you like to open the following link in your browser?
+                            """.L10N("Client:Main:OpenLinkConfirmationText")
+                            + Environment.NewLine + Environment.NewLine + link,
+                            XNAMessageBoxButtons.YesNo);
+                        msgBox.YesClickedAction = (msgBox) => ProcessLauncher.StartShellProcess(link);
+                        msgBox.Show();
```

Comment:
I think this should be made into a function or something because it repeats

Reply:
> No, because this is the only second place where it used.

I mean, that's exactly the point when such an amount of lines/code should be split out into a function. Imagine someone adjusting one and forgetting to adjust another.